lib/Mojolicious/Guides/Testing.pod

```diff
@@ -0,0 +1,814 @@
+us. The next C<send_ok> assertion tests the connection again (in case it
+closed, for example) and attempts to send the message C<Hello Mojo!>. The next
+assertion, C<message_ok>, blocks (using the L<Mojo::IOLoop> singleton in the
+application) and waits for a response from the server. The response is then
+compared with C<'echo: Hello Mojo!'> in the C<message_is> assertion, and
+finally we close and test our connection status again with C<finish_ok>.
+
+The second test is like the first, but now we're sending and expecting JSON
+documents at C</echo.json>. In the C<send_ok> assertion we take advantage of
+L<Mojo::UserAgent>'s JSON content generator (see
+L<Mojo::UserAgent::Transactor>) to marshal hash and array references into JSON
+documents, and then send them as a WebSocket message. We wait (block) for a
+response from the server with C<message_ok>. Then because we're expecting a
+JSON document back, we can leverage C<json_message_ok> which parses the
+WebSocket response body and returns an object we can access through
+L<Mojo::JSON::Pointer> syntax. Then we close (and test) our WebSocket
+connection.
+
+Testing WebSocket servers does not get any simpler than with L<Test::Mojo>.
+
+=head2 Extending L<Test::Mojo>
+
+In some situations you may find your tests have a lot of boilerplate, making
+tests cluttered and hard to follow. For example let's say we have an RPC-based
+web application where the HTTP method is always C<POST> and the remote
+procedure name is in the URL, and we use HTTP Basic authentication with
+username C<me> and password C<secretly>.
+
+All of our requests follow this basic pattern:
+
+  POST /<service>/<command>
+  Authorization: Basic xxxxx
+
+  {"json":"payload"}
+
+For example:
+
+  POST /bedroom-light/power-on
+  Authorization: Basic bWU6c2VjcmV0bHkK=
+
+  {"level":"50%"}
+
+Or:
+
+  POST /blender/spin-blades
+  Authorization: Basic bWU6c2VjcmV0bHkK=
+
+  {"speed":"whip"}
+
+Or:
+
+  POST /thermostat/heat
+  Authorization: Basic bWU6c2VjcmV0bHkK=
+
+  {"temperature":"23°C"}
+
+To test this kind of web application, we'd need to do something like this for
+every test:
+
+  $t->post_ok('/thermostat/heat',
+              {Authorization => 'Basic ' . b64_encode 'me:secretly', ''},
+              json => {temperature => '23°C'})
+    ->status_is(200);
+
+But adding the C<Authorization> header for every request is tedious and likely
+to result in copy/paste errors given enough tests. We'd like to extend
+C<post_ok>'s behavior to build the header for us. With Mojolicious 7.41 (or
+later) and L<Role::Tiny> installed, we can compose this behavior with roles.
+
+First, we create a new package that will implement the role we want:
+
+  package Test::Mojo::Role::BasicAuth;
+  use Role::Tiny;
+  use Mojo::Util 'b64_encode';
+  use Mojo::Base 'Test::Mojo';
+
+  sub post_with_auth_ok {
+      my $t = shift;
+      my $url = shift;
+      my $creds = shift; # username:password
```

Comment:
And i don't see a reason not to write `my ($t, $url, $creds) = (shift, shift, shift);`.